lib/ElmJson.js

```diff
@@ -0,0 +1,193 @@
+    'the file'
+  );
+
+  switch (json.type) {
+    case 'application':
+      return {
+        ...json,
+        type: 'application',
+        'source-directories': parseSourceDirectories(
+          json['source-directories']
+        ),
+        dependencies: parseDirectAndIndirectDependencies(
+          json.dependencies,
+          'dependencies'
+        ),
+        'test-dependencies': parseDirectAndIndirectDependencies(
+          json['test-dependencies'],
+          'test-dependencies'
+        ),
+      };
+
+    case 'package':
+      return {
+        ...json,
+        type: 'package',
+        dependencies: parseDependencies(json.dependencies, 'dependencies'),
+        'test-dependencies': parseDependencies(
+          json['test-dependencies'],
+          'test-dependencies'
+        ),
+      };
+
+    default:
+      throw new Error(
+        `Expected "type" to be "application" or "package", but got: ${stringify(
+          json.type
+        )}`
+      );
+  }
+}
+
+function parseSourceDirectories(json /*: mixed */) /*: Array<string> */ {
+  if (!Array.isArray(json)) {
+    throw new Error(
+      `Expected "source-directories" to be an array, but got: ${stringify(
+        json
+      )}`
+    );
+  }
+
+  const result = [];
+
+  for (const [index, item] of json.entries()) {
+    if (typeof item !== 'string') {
+      throw new Error(
+        `Expected "source-directories"->${index} to be a string, but got: ${stringify(
+          item
+        )}`
+      );
+    }
+    result.push(item);
+  }
+
+  if (result.length === 0) {
+    throw new Error(
+      'Expected "source-directories" to contain at least one item, but it is empty.'
+    );
+  }
+
+  return result;
+}
+
+function parseDirectAndIndirectDependencies(
+  json /*: mixed */,
+  what /*: string */
+) /*: typeof DirectAndIndirectDependencies */ {
+  const jsonObject = parseObject(json, what);
+  return {
+    direct: parseDependencies(jsonObject.direct, `${what}->"direct"`),
+    indirect: parseDependencies(jsonObject.indirect, `${what}->"indirect"`),
+  };
+}
+
+function parseDependencies(
+  json /*: mixed */,
+  what /*: string */
+) /*: typeof Dependencies */ {
+  const jsonObject = parseObject(json, what);
+  const result = {};
+
+  for (const [key, value] of Object.entries(jsonObject)) {
+    if (typeof value !== 'string') {
+      throw new Error(
+        `Expected ${what}->${stringify(
+          key
+        )} to be a string, but got: ${stringify(value)}`
+      );
+    }
+    result[key] = value;
+  }
+
+  return result;
+}
+
+function parseObject(
+  json /*: mixed */,
+  what /*: string */
+) /*: { +[string]: mixed } */ {
+  if (json == null || typeof json !== 'object' || Array.isArray(json)) {
+    throw new Error(
+      `Expected ${what} to be an object, but got: ${stringify(json)}`
+    );
+  }
+  return json;
+}
+
+function stringify(json /*: mixed */) /*: string */ {
+  const maybeString = JSON.stringify(json);
```

Comment:
Minor comment: `JSON.stringify` can throw a `TypeError`

Reply:
True! However, not in this particular file. The only things I’m ever stringifying are things that just came from a `JSON.parse`.